### DISCUSSION TITLE
Proposal for PixiJS Ecosystem of Plugins

### DISCUSSION THREAD

**Discussion by bigtimebuddy** [pixijs/pixijs#7056] [Ideas]
There been confusion in the PixiJS community around the ecosystem of plugins. Not all plugins are of the same quality, some are not updated or do not work with the latest version of PixiJS. If you're a developer that is looking to extend PixiJS functionality, it's hard to know what's official vs. non-official vs. semi-official or what's stable, production, experimental. 

Let's break down some examples from the *pixijs* organization on GitHub. I don't know all this information, so this is mostly just a guess. **Disclaimer: This might not be accurate!**

| Repo | Primary Contributor | Maturity | Active | v5 Compatible |
| --- | --- | --- | --- | -- |
| pixi-projection | @ivanpopelyshev  | Experimental? | ❔   | ✅  |
| pixi-layers | @ivanpopelyshev  | Stable | ❔   | ✅  |
| pixi-spine | @ivanpopelyshev  | Stable | ❔   | ✅  |
| pixi-ui | @SukantPal |  Experimental | ❔    | ✅  |
| pixi-animate | @andrewstart  | Stable (v1), Experimental (v2) | ✅  | ✅  |
| pixi-sound | @bigtimebuddy  | Stable | Stale  | ✅  |
| pixi-particles | @andrewstart | Stable | ✅  | ✅  |
| pixi-tilemap | @ivanpopelyshev | Stable| ❔   | ✅   |
| pixi-picture | @ivanpopelyshev | Experimental| ❔  | ✅   |
| pixi-html-text | @bigtimebuddy  | Experimental | Stale  | ✅  | 
| pixi-batch-renderer | @SukantPal  | Stable | ✅   | ✅  | 
| pixi-lights | @ivanpopelyshev |Experimental |❔  | ❔  |
| pixi-filters | @bigtimebuddy | Stable | ✅  | ✅  |
| pixi-compressed-textures | @ivanpopelyshev | Deprecated? | ❔  | ✅  | 

It's a little all over the map. Most of these are "official" in the sense that they live on the pixijs org, but these are often passion projects of a single core team member. These plugins are down-stream from **pixi.js** and are not always update-to-date with the latest. Some use private APIs which make them brittle to compatibility with the main project.

* If you are one of the contributors to the above project, could you help fill in some missing information here?
* Any ideas how do we clarify this mess for developers wanting to use these plugins?

### Proposal

I would like to see all plugins meet some basic criteria so that we can give it the PixiJS seal of approval. This will signal to users that these plugins obey a few rules so at least there's confidence.

- [ ] Ships TypeScript typings
- [ ] Provides a build for browser
- [ ] Provides a build(s) for bundlers (Webpack)
- [ ] Does not use private APIs (which are brittle)
- [ ] Uses peerDependencies to properly define compatibility
- [ ] Uses base-packages for peerDependencies (e.g. `@pixi/core`, not `pixi.js` or `pixi.js-legacy`)
- [ ] Respects Semantic Versioning
- [ ] Works with the latest major version

**Comment by ShukantPal** [comment]
Regarding pixi-batch-renderer, it is active in the sense that it will get bug fixes proactively.

✅ Ships TypeScript typings
❌ Does not provide a build for browser
✅ Provide build(s) for bundles
✅ No black magic with internal core APIs
❌ Uses pixi.js as a dependency
✅ Respects semantic versioning
✅ Works with latest major version

Note that pixi-batch-renderer is both a monorepo and a package (the package itself is not the monorepo :P)

Edit: pixi-batch-renderer is **not** experimental. It is stable and is used at by one of my clients to combine multiple shaders into one batch shader on very large scene graphs (including those generated by 10<sup>3</sup>-dimension SVGs rendered by @pixi-essentials/svg).

**Reply by bigtimebuddy** [reply]
Updated table with this information. Thanks for clarifying

**Comment by ShukantPal** [comment]
In order to help unify these PixiJS plugins, I’d like to propose formalizing a proper documentation site for all projects that would look something like sukantpal.github.io/pixi-api. (The site is in a very, very early stage).

We should probably include proper API documentation for all plugins @bigtimebuddy.
